// @jsoverson/shift-codegen/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = codeGen;
Object.defineProperty(exports, "MinimalCodeGen", {
  enumerable: true,
  get: function () {
    return _minimalCodegen.default;
  }
});
Object.defineProperty(exports, "ExtensibleCodeGen", {
  enumerable: true,
  get: function () {
    return _formattedCodegen.ExtensibleCodeGen;
  }
});
Object.defineProperty(exports, "FormattedCodeGen", {
  enumerable: true,
  get: function () {
    return _formattedCodegen.FormattedCodeGen;
  }
});
Object.defineProperty(exports, "Sep", {
  enumerable: true,
  get: function () {
    return _formattedCodegen.Sep;
  }
});
Object.defineProperty(exports, "Precedence", {
  enumerable: true,
  get: function () {
    return _coderep.Precedence;
  }
});
Object.defineProperty(exports, "getPrecedence", {
  enumerable: true,
  get: function () {
    return _coderep.getPrecedence;
  }
});
Object.defineProperty(exports, "escapeStringLiteral", {
  enumerable: true,
  get: function () {
    return _coderep.escapeStringLiteral;
  }
});
Object.defineProperty(exports, "CodeRep", {
  enumerable: true,
  get: function () {
    return _coderep.CodeRep;
  }
});
Object.defineProperty(exports, "Empty", {
  enumerable: true,
  get: function () {
    return _coderep.Empty;
  }
});
Object.defineProperty(exports, "Token", {
  enumerable: true,
  get: function () {
    return _coderep.Token;
  }
});
Object.defineProperty(exports, "NumberCodeRep", {
  enumerable: true,
  get: function () {
    return _coderep.NumberCodeRep;
  }
});
Object.defineProperty(exports, "Paren", {
  enumerable: true,
  get: function () {
    return _coderep.Paren;
  }
});
Object.defineProperty(exports, "Bracket", {
  enumerable: true,
  get: function () {
    return _coderep.Bracket;
  }
});
Object.defineProperty(exports, "Brace", {
  enumerable: true,
  get: function () {
    return _coderep.Brace;
  }
});
Object.defineProperty(exports, "NoIn", {
  enumerable: true,
  get: function () {
    return _coderep.NoIn;
  }
});
Object.defineProperty(exports, "ContainsIn", {
  enumerable: true,
  get: function () {
    return _coderep.ContainsIn;
  }
});
Object.defineProperty(exports, "Seq", {
  enumerable: true,
  get: function () {
    return _coderep.Seq;
  }
});
Object.defineProperty(exports, "Semi", {
  enumerable: true,
  get: function () {
    return _coderep.Semi;
  }
});
Object.defineProperty(exports, "CommaSep", {
  enumerable: true,
  get: function () {
    return _coderep.CommaSep;
  }
});
Object.defineProperty(exports, "SemiOp", {
  enumerable: true,
  get: function () {
    return _coderep.SemiOp;
  }
});
Object.defineProperty(exports, "codeGenWithLocation", {
  enumerable: true,
  get: function () {
    return _withLocation.default;
  }
});

var _shiftReducer = _interopRequireDefault(require("shift-reducer"));

var _tokenStream = require("./token-stream");

var _minimalCodegen = _interopRequireDefault(require("./minimal-codegen"));

var _formattedCodegen = require("./formatted-codegen");

var _coderep = require("./coderep");

var _withLocation = _interopRequireDefault(require("./with-location"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function codeGen(script, generator = new _minimalCodegen.default()) {
  let ts = new _tokenStream.TokenStream();
  let rep = (0, _shiftReducer.default)(generator, script);
  rep.emit(ts);
  return ts.result;
}
